File: backend/servo.py
#!/usr/bin/python3

import sys
import RPi.GPIO as GPIO
from time import sleep

def go(port, angle, delay):
    GPIO.setmode(GPIO.BOARD)
    GPIO.setup(port, GPIO.OUT)
    GPIO.output(port, GPIO.LOW)
    sleep(0.1)
    f = 100
    p = GPIO.PWM(port, f)
    d = angle/180*(2.38-0.55)+0.55
    if angle < 1:
        d = 0.50
    elif angle > 179:
        d = 2.39
    dc = d*f/10
    p.start(dc)
    sleep(delay)
    p.stop()
    GPIO.cleanup()
    sleep(delay)

port = int(sys.argv[1])
angle = float(sys.argv[2])

go(port, angle, 1.00)
go(port, angle, 0.50)
go(port, angle, 0.25)

File: middleware/io/servo.cpp
#include "io/servo.hpp"
#include <thread>
#include "io/shared.hpp"

using namespace std::chrono_literals;

servo::servo(int pin) : _pin{ pin } { }

#pragma clang diagnostic push
#pragma ide diagnostic ignored "bugprone-unused-raii"
sink<1> &servo::operator<<(const arr_t<1> &r) {
    auto d{ r[0]/180*(2.38-0.55)+0.55 }; // duration in ms
    if (r[0] < 1)
        d = 0.50;
    else if (r[0] > 179)
        d = 2.39;
    auto dc{ d*freq/10 };

    _th = std::thread{[this, dc](){
        lock_shared l{ g_pwm };
        {
            py{} << R"py(
import RPi.GPIO as GPIO
from time import sleep
port = )py" << _pin << R"py(
GPIO.setmode(GPIO.BOARD)
GPIO.setup(port, GPIO.OUT)
p = GPIO.PWM(port, )py" << freq << R"py()
p.start()py" << dc << R"py()
sleep(0.4)
p.stop()
sleep(0.1)
p.start()py" << dc << R"py()
sleep(0.4)
p.stop()
GPIO.cleanup()
)py";
        }
        std::this_thread::sleep_for(1.5s);
    }};
    return *this;
}
#pragma clang diagnostic pop
